packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts

```diff
@@ -313,6 +329,79 @@ export default class UserStorageController extends BaseController<
     isNetworkSyncingEnabled: false,
   };
 
+  #_snapPublicKeyCache: string | null = null;
+
+  #keyWrapping = {
+    /**
+     * Returns the snap Encryption public key.
+     *
+     * @returns The snap Encryption public key.
+     */
+    snapGetEncryptionPublicKey: async (): Promise<string> => {
+      if (this.#_snapPublicKeyCache) {
+        return this.#_snapPublicKeyCache;
+      }
+
+      if (!this.#isUnlocked) {
+        throw new Error(
+          '#snapGetEncryptionPublicKey - unable to call snap, wallet is locked',
+        );
+      }
+
+      const result = (await this.messagingSystem.call(
+        'SnapController:handleRequest',
+        createSnapEncryptionPublicKeyRequest(),
+      )) as string;
+
+      this.#_snapPublicKeyCache = result;
+
+      return result;
+    },
+
+    /**
+     * Decrypts a message using the message signing snap.
+     *
+     * @param data - Eip1024EncryptedData - The encrypted data.
+     * @returns The decrypted message, if it was intended for this wallet, null otherwise. TODO: check error scenarios
+     */
+    snapDecryptMessage: async (data: Eip1024EncryptedData): Promise<string> => {
+      if (!this.#isUnlocked) {
+        throw new Error(
+          '#snapDecryptMessage - unable to call snap, wallet is locked',
+        );
+      }
+
+      const result = (await this.messagingSystem.call(
+        'SnapController:handleRequest',
+        createSnapDecryptMessageRequest(data),
+      )) as string;
+
+      return result;
+    },
+
+    loadWrappedKey: async (keyRef: string): Promise<string | null> => {
+      return this.state.encryptedContentKeys[keyRef] ?? null;
+    },
+
+    storeWrappedKey: (keyRef: string, wrappedKey: string): Promise<void> => {
+      return new Promise((resolve) => {
+        this.update((state) => {
+          state.encryptedContentKeys[keyRef] = wrappedKey;
+          resolve();
+        });
+      });
+    },
```

Comment:
Do these need to be promises?

If we do want to make then pseudo promises, we can make the functions async and avoid return explicit promises.

```ts
// By using the `async` function syntax, the return type will be inferred as a promise.
storeWrappedKey: async (...): Promise<void> => {
  this.update(...)
}
```